### PR TITLE
fix issue 20324 - Calling __traits(getUnitTests) on a template causes compiler segfault

### PR DESCRIPTION
**Pull request by ghost**
The idea is that you cannot get the tests of a partially specialized type. unittest blocks inside a partially specialized type may use or not the template parameters so allowing them would require a complex analysis.

### CODE REVIEW

File: test/fail_compilation/ice20324.d [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+template Foo() { unittest{} }
+class Bar(T)   { unittest{} }
+
+// template decls rejected
+alias fooTests = __traits(getUnitTests, Foo);
+alias barTests = __traits(getUnitTests, Bar);
+
+// template instances accepted
+static assert(__traits(getUnitTests, Foo!()).length == 1);
```
Comment (clYd3r) [inline test/fail_compilation/ice20324.d]: This should be in a "compiles" test, for some reason it passes locally when testing the single file but not when running the whole test suite.

Comment (clYd3r) [reply test/fail_compilation/ice20324.d]: Mmmmh just the -unittest switch missing actually

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
It would be better to say what the type passed is.

**Comment by clYd3r** [comment]
I have doubts anyway. I used a tester based on the trait but my own code doesn't contain any inner unittest block so I need to verify one or two things. TLDR; don't merge fast... there's two or three weeks because next bug fix release anyway.

**Comment by Geod24** [comment]
Applied a label, feel free to ping to have it removed.

**Comment by clYd3r** [comment]
I've changed the  approach and decided not to issue an error.

**Comment by clYd3r** [comment]
ping for the removal of the blocked "label", checks done.

**Comment by Geod24** [comment]
> I've changed the approach and decided not to issue an error.

That might be a surprising behavior. Can you expand on why you went with this approach ? Adding the explanation in the commit message would be great too.

**Comment by clYd3r** [comment]
1. This makes writing a generic tester more friendly, with less `static if`.
2. On the principle of least astonishment, I'm less astonished by no error than by an error. It's pretty obvious that templates that are not instances are not usable. For example when I made the fix I didn't even tried to see where the AV happened, I directly opened traits.d because the place to fix the crash was obvious.

**Comment by RazvanN7** [comment]
Hmm, I wanted to close and reopen to restart the autotester, but it seems like the branch was deleted and I cannot reopen. It seems like the github account was deleted.

**Pull request closed without merging**